### ISSUE TITLE
`let_underscore_untyped` doesn't get triggered when `_` is used without `let`

### ISSUE THREAD

**Issue by zacknewman** [rust-lang/rust-clippy#15658]
### Summary

The justification for the `let_underscore_untyped` lint seems to apply when `_` is used without a `let` binding. I realize this means the use of `_` is impossible with this lint, but that's not uncommon for `restriction`lints.

```rust
#![expect(dead_code, reason = "illustrative example")]
#![deny(clippy::let_underscore_untyped, reason = "illustrative example")]
fn bar() {
    // Doesn't this suffer from the same drawbacks as `let _ = foo();`?
    _ = foo();
    // Uncommenting below will cause a compile error due to the `let_underscore_untyped` lint.
    // let _ = foo();
}
fn foo() -> Result<(), u32> {
    Ok(())
}
```

**Comment by kumarUjjawal** [comment]
My take is `_ = expr` is widely used as the “I am intentionally discarding this” form. Clippy’s current `let_underscore_untyped` explicitly suggests it as an alternative to untyped `let _ = …`, so flagging it would contradict the lint’s own guidance and break existing workflows.

**Comment by zacknewman** [comment]
> My take is `_ = expr` is widely used as the “I am intentionally discarding this” form. Clippy’s current `let_underscore_untyped` explicitly suggests it as an alternative to untyped `let _ = …`, so flagging it would contradict the lint’s own guidance and break existing workflows.

I don't agree that with take, and what you claim about Clippy's suggestion is wrong. `let _` and `_` are semantically equivalent and mean "I am intentionally discarding this". The only difference is that `let _` is older—it's worked since `rustc 1.0.0`—but assignment to `_` was introduced in Rust 1.59 as part of the ability for destructuring assignments. Additionally the compiler and Clippy recommend `let _`, _not_ `_`, for the below code:

```rust
#![deny(
    unused_must_use,
    clippy::let_underscore_untyped,
    reason = "illustrative example"
)]
#![expect(dead_code, reason = "illustrative example")]
const fn bar() {
    foo();
}
const fn foo() -> Result<(), u32> {
    Ok(())
}
```

`cargo check` and `cargo clippy` emit the below suggestion:


```bash
error: unused `Result` that must be used
 --> src/lib.rs:8:5
  |
8 |     foo();
  |     ^^^^^
  |
  = note: this `Result` may be an `Err` variant, which should be handled
  = note: illustrative example
note: the lint level is defined here
 --> src/lib.rs:2:5
  |
2 |     unused_must_use,
  |     ^^^^^^^^^^^^^^^
help: use `let _ = ...` to ignore the resulting value
  |
8 |     let _ = foo();
  |     +++++++

error: could not compile `example` (lib) due to 1 previous error
```

If you change the code to the below:

```rust
#![deny(
    unused_must_use,
    clippy::let_underscore_untyped,
    reason = "illustrative example"
)]
#![expect(dead_code, reason = "illustrative example")]
const fn bar() {
    let _ = foo();
}
const fn foo() -> Result<(), u32> {
    Ok(())
}
```

then `cargo clippy` emits the suggestion to add a type annotation, _not_ to use `_` as you claim:

```bash
error: non-binding `let` without a type annotation
 --> src/lib.rs:8:5
  |
8 |     let _ = foo();
  |     ^^^^^^^^^^^^^^
  |
help: consider adding a type annotation
 --> src/lib.rs:8:10
  |
8 |     let _ = foo();
  |          ^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#let_underscore_untyped
  = note: illustrative example
note: the lint level is defined here
 --> src/lib.rs:3:5
  |
3 |     clippy::let_underscore_untyped,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not compile `example` (lib) due to 1 previous error
```

Regardless, the justification for this lint is that subtle bugs may be introduced if the type is changed which applies to both `_` and `let _`. Of course the literal name of the lint does clearly suggest that `let` is required, so I guess one could argue for a _separate_ lint using the same exact justification as this lint but relax the requirement that `let` be used and simply apply to _any_ discarded value.

**Comment by samueltardieu** [comment]
`let _ = …` and `_ = …` also behave differently regarding temporary lifetime extension, so they are not always semantically equivalent.

**Comment by zacknewman** [comment]
> `let _ = …` and `_ = …` also behave differently regarding temporary lifetime extension, so they are not always semantically equivalent.

Fair, but my stance remains the same unless one can provide even _one_ example where the use of `_` is not affected by these same "subtle bugs" that are referenced to by the justification for this lint. More explicitly, I'd like to see an example where

1. `let _` (_without_ the type annotation) is affected by "subtle bugs" when the type changes.
2. `let _: <type>` is not affected by these same "subtle bugs" when the type changes (i.e., prove this lint's utility).
3. `_` is not affected by the same "subtle bugs" when the type changes.

If not, then either `_` should be included in this lint's analysis or there should be another lint that only applies to `_` or applies to both `_` and `let _` that way these same "subtle bugs" are avoided for `_` as well.

Based on your response, if such an example exists, it would seem to rely on the differences of how temporary lifetime extensions are applied. I didn't realize this lint was really only about temporary lifetime extensions. I thought the "subtle bugs" that are referenced applied to a variety of types. I'd still appreciate an example though.

The reason I enable this lint is that sometimes I want to discard values of certain types but not others. For example `let _: bool = foo();` allows me to discard the `bool` returned from `foo`; however if `foo` were to return `Result<T, E>`, then I don't want to discard it. `let: bool` prevents that from happening. If I were to instead use `_ = foo();`, then suddenly I'd discard a `Result<T, E>` without knowing suggesting that this lint should apply to `_` as well.

**Comment by kumarUjjawal** [comment]
@zacknewman You’re right my earlier analysis was off.  As you said we can either;
( 1) extend `let_underscore_untyped` to cover _ = expr; (with a clearer message), or 
(2) add a separate lint for `_ = expr`.

@samueltardieu @zacknewman what would be the best path forward?

**Comment by zacknewman** [comment]
> [@zacknewman](https://github.com/zacknewman) You’re right my earlier analysis was off. As you said we can either; ( 1) extend `let_underscore_untyped` to cover _ = expr; (with a clearer message), or (2) add a separate lint for `_ = expr`.
> 
> [@samueltardieu](https://github.com/samueltardieu) [@zacknewman](https://github.com/zacknewman) what would be the best path forward?

I think this lint should cover `_`, and I think the `I-false_negative` tag should be added back _unless_ there are "subtle bugs" that affect `let _` (without the type annotation) that don't affect `_`. In which case, then a new lint should be added.

I claim that the _only_ class of "subtle bugs" that this lint purports to prevent are the ones that I mentioned above. Namely to prevent discarding values of different types. Just because I want to discard a `bool` today doesn't mean I want to discard a `Result` tomorrow. If one uses `_` or `let _`, they will happily discard the `Result`. This means people that want to use this lint are not actually preventing the bugs they want; thus making this a false negative and warrants a change.

This lint is both `allow`-by-default _and_ part of the `restriction` group thus making this change more appropriate. If people are suddenly impacted by a bunch of warnings or errors due to this change, then that's a good thing. If they don't want to address it, then they never wanted to use this lint to begin with and should simply make it `allow` again. This lint is likely not used by many people; and the ones that do use it, probably don't enable the entire `restriction` group—`blanket_clippy_restriction_lints` is `warn`-by-default to discourage this. This means those that actually want to use this lint won't benefit from this fix; and since they likely don't enable the entire `restriction` group, they won't benefit from the new lint either unless they keep up with new lints and know to also enable it.

**Comment by samueltardieu** [comment]
This is not a false negative in the sense where the lint description clearly tells that this lints `let _` constructs. Clearly, `_` was not considered as a problem when the lint was written, and moreover it was suggested to use `_` instead of `let _` (which is wrong IMO).

So you want to extend the lint to `_` which is fine, so this is a feature request, and a discussion is needed to determine whether this would be an extension of this lint or a new lint. It also needs to warn when this might cause the drop order to be affected: `_ = &return_something_with_drop();` will drop the value returned by the function at the end of the statement, while `let _: SomeType = &return_something_with_drop();` will drop it at the end of the enclosing scope. Fortunately, in this direction it should not be encountered as adding the `&` is silly in the first place (while it is not in a `let _`).

**Comment by zacknewman** [comment]
> This is not a false negative in the sense where the lint description clearly tells that this lints `let _` constructs. Clearly, `_` was not considered as a problem when the lint was written, and moreover it was suggested to use `_` instead of `let _` (which is wrong IMO).

Fine. What is done is done, and I can't travel back in time to argue that the logic that is used to justify this lint's (legitimate) existence also applies to `_`. I just want to prevent myself from writing code that allows me to discard things I didn't intend when the return type of a function is changed.

> So you want to extend the lint to `_` which is fine, so this is a feature request, and a discussion is needed to determine whether this would be an extension of this lint or a new lint. It also needs to warn when this might cause the drop order to be affected: `_ = &return_something_with_drop();` will drop the value returned by the function at the end of the statement, while `let _: SomeType = &return_something_with_drop();` will drop it at the end of the enclosing scope. Fortunately, in this direction it should not be encountered as adding the `&` is silly in the first place (while it is not in a `let _`).

Hm, because drop order is affected, I'm ambivalent if this should be a new lint or not. Selfishly, I'll benefit if a new lint is added since I raised this issue; additionally, I'm one of those "weirdos" that `deny`s the `restriction` group (and `allow`s `blanket_clippy_restriction_lints`); so even if I hadn't been aware of this issue, the new lint would be automatically enabled. I was more looking out for the others who won't benefit because they don't enable the `restriction` group nor keep up with new `restriction` lints to see which ones they want.

**Comment by samueltardieu** [comment]
> Hm, because drop order is affected, I'm ambivalent if this should be a new lint or not

Note that this is not a problem *per-se*: a note can be added, and if a fix is suggested it might be marked as "maybe incorrect" to prevent it from being blindly applied.

**Comment by zacknewman** [comment]
> Note that this is not a problem _per-se_: a note can be added, and if a fix is suggested it might be marked as "maybe incorrect" to prevent it from being blindly applied.

Your example, no matter how contrived, is enough to quell my insistence that this change be added to this lint though. To me, this lint is entirely about avoiding the _future_ semantics from being affected by a discard and not the semantics at the time the code was written. You have shown that the semantics of the _current_ code is affected due to the (possible) change in drop order.

There is also the human element. I'm not a maintainer, and unfortunately I could see issues being raised where this new behavior is not desired no matter how vehemently I think the logic applies to `_` as well. So making this a new lint should hopefully prevent that—_technically_ you could argue that not making the change to this lint could also cause new issues where people insist the new lint's ability be merged into this lint, but I think that's probably less likely.

Anyway, thanks for hearing me out.
